File: library/src/main/java/com/qwert2603/permesso/internal/ActivityProvider.java
package com.qwert2603.permesso.internal;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;

import io.reactivex.Single;
import io.reactivex.functions.Function;
import io.reactivex.functions.Predicate;
import io.reactivex.subjects.BehaviorSubject;

final class ActivityProvider {

    private static final class Wrapper<T> {
        @Nullable
        final T value;

        Wrapper(@Nullable T value) {
            this.value = value;
        }
    }

    static final ActivityProvider INSTANCE = new ActivityProvider();

    private final BehaviorSubject<Wrapper<AppCompatActivity>> activityChanges = BehaviorSubject.createDefault(new Wrapper<AppCompatActivity>(null));

    private ActivityProvider() {
    }

    @NonNull
    Single<AppCompatActivity> resumedActivity() {
        return activityChanges
                .filter(new Predicate<Wrapper<AppCompatActivity>>() {
                    @Override
                    public boolean test(Wrapper<AppCompatActivity> activityWrapper) {
                        return activityWrapper.value != null;
                    }
                })
                .map(new Function<Wrapper<AppCompatActivity>, AppCompatActivity>() {
                    @Override
                    public AppCompatActivity apply(Wrapper<AppCompatActivity> activityWrapper) {
                        return activityWrapper.value;
                    }
                })
                .firstOrError();
    }

    void onActivityResumed(AppCompatActivity activity) {
        activityChanges.onNext(new Wrapper<>(activity));
    }

    void onActivityPaused(@SuppressWarnings("unused") AppCompatActivity activity) {
        activityChanges.onNext(new Wrapper<AppCompatActivity>(null));
    }
}

File: permesso_coroutines/src/main/java/com/qwert2603/permesso_coroutines/ActivityProvider.kt
package com.qwert2603.permesso_coroutines

import androidx.appcompat.app.AppCompatActivity
import kotlinx.coroutines.channels.ConflatedBroadcastChannel
import kotlinx.coroutines.channels.sendBlocking

internal object ActivityProvider {
    data class Wrapper<T>(val value: T?)

    val activityChanges = ConflatedBroadcastChannel<Wrapper<AppCompatActivity>>()

    fun onActivityResumed(activity: AppCompatActivity) {
        activityChanges.sendBlocking(Wrapper(activity))
    }

    fun onActivityPaused(activity: AppCompatActivity) {
        activityChanges.sendBlocking(Wrapper<AppCompatActivity>(null))
    }
}